src/spotify_webapi.c

```diff
@@ -0,0 +1,754 @@
+	  (keyval_add(&kv, "client_secret", spotify_client_secret) == 0) &&
+          (keyval_add(&kv, "refresh_token", refresh_token) == 0) );
+  if (!ret)
+    {
+      DPRINTF(E_LOG, L_SPOTIFY, "Add parameters to keyval failed");
+      ret = -1;
+    }
+  else
+    ret = tokens_get(&kv, &err);
+
+  free(refresh_token);
+  keyval_clear(&kv);
+
+  return ret;
+}
+
+static int
+request_uri(struct spotify_request *request, const char *uri)
+{
+  char bearer_token[1024];
+  int ret;
+
+  memset(request, 0, sizeof(struct spotify_request));
+
+  if (0 > spotifywebapi_token_refresh())
+    {
+      return -1;
+    }
+
+  request->ctx = calloc(1, sizeof(struct http_client_ctx));
+  request->ctx->output_headers = calloc(1, sizeof(struct keyval));
+  request->ctx->input_body = evbuffer_new();
+  request->ctx->url = uri;
+
+  snprintf(bearer_token, sizeof(bearer_token), "Bearer %s", spotify_access_token);
+  if (keyval_add(request->ctx->output_headers, "Authorization", bearer_token) < 0)
+    {
+      DPRINTF(E_LOG, L_SPOTIFY, "Add bearer_token to keyval failed\n");
+      return -1;
+    }
+
+  ret = http_client_request(request->ctx);
+  if (ret < 0)
+    {
+      DPRINTF(E_LOG, L_SPOTIFY, "Request for saved tracks/albums failed\n");
+      return -1;
+    }
+
+  // 0-terminate for safety
+  evbuffer_add(request->ctx->input_body, "", 1);
+
+  request->response_body = (char *) evbuffer_pullup(request->ctx->input_body, -1);
+  if (!request->response_body || (strlen(request->response_body) == 0))
+    {
+      DPRINTF(E_LOG, L_SPOTIFY, "Request for saved tracks/albums failed, response was empty\n");
+      return -1;
+    }
+
+//  DPRINTF(E_DBG, L_SPOTIFY, "Wep api response for '%s'\n%s\n", uri, request->response_body);
+
+  request->haystack = json_tokener_parse(request->response_body);
+  if (!request->haystack)
+    {
+      DPRINTF(E_LOG, L_SPOTIFY, "JSON parser returned an error\n");
+      return -1;
+    }
+
+  DPRINTF(E_DBG, L_SPOTIFY, "Got response for '%s'\n", uri);
+  return 0;
+}
+
+void
+spotifywebapi_request_end(struct spotify_request *request)
+{
+  http_client_ctx_free(request->ctx);
+  jparse_free(request->haystack);
+}
+
+int
+spotifywebapi_request_next(struct spotify_request *request, const char *uri)
+{
+  char *next_uri;
+  const char *tmp;
+  int ret;
+
+  if (request->ctx && !request->next_uri)
+    {
+      // Reached end of paging requests, terminate loop
+      return -1;
+    }
+
+  if (!request->ctx)
+    {
+      // First paging request
+      next_uri = strdup (uri);
+    }
+  else
+    {
+      // Next paging request
+      next_uri = strdup(request->next_uri);
+      spotifywebapi_request_end(request);
+    }
+
+  ret = request_uri(request, next_uri);
+  free(next_uri);
+
+  if (ret < 0)
+    return ret;
+
+  request->total = jparse_int_from_obj(request->haystack, "total");
+  tmp = jparse_str_from_obj(request->haystack, "next");
+  if (tmp)
+    request->next_uri = strdup(tmp);
```

Comment:
Does this mean that if there is no tmp (next is missing), then ->next_uri will be unchanged and it will continue requesting the same page forever? Please safeguard against "ddos'ing" Spotify :-)